filebeat/input/filestream/takeover/takeover.go

```diff
@@ -0,0 +1,237 @@
+	}
+
+	statesToSet, statesToRemove, err := takeOverStates(log, store, filestreamMatchers)
+	if err != nil {
+		return fmt.Errorf("failed to take over one of loginput states: %w", err)
+	}
+	if len(statesToSet) == 0 {
+		log.Info("no state to take over")
+		return nil
+	}
+
+	// before making changes, we backup the registry files for manual rollback if needed
+	err = backuper.Backup()
+	if err != nil {
+		return fmt.Errorf("failed to create backup files: %w", err)
+	}
+
+	for key := range statesToSet {
+		err = store.Set(key, statesToSet[key])
+		if err != nil {
+			return fmt.Errorf("failed to set the taken state: %w", err)
+		}
+	}
+	for key := range statesToRemove {
+		err = store.Remove(key)
+		if err != nil {
+			return fmt.Errorf("failed to remove the taken state: %w", err)
+		}
+	}
+
+	log.Infof("filestream inputs took over %d file(s) from loginputs", len(statesToSet))
+	return nil
+}
+
+func takeOverStates(log *logp.Logger, store backend.Store, matchers filestreamMatchers) (toSet map[string]mapstr.M, toRemove map[string]struct{}, err error) {
+	toSet = make(map[string]mapstr.M)
+	toRemove = make(map[string]struct{})
+
+	err = store.Each(func(key string, value statestore.ValueDecoder) (bool, error) {
+		if !strings.HasPrefix(key, loginputPrefix) {
+			return true, nil
+		}
+		state := make(mapstr.M)
+		err := value.Decode(&state)
+		if err != nil {
+			return false, err
+		}
+		sourceValue, err := state.GetValue("source")
+		if errors.Is(err, mapstr.ErrKeyNotFound) {
+			return true, nil
+		}
+		if err != nil {
+			return false, fmt.Errorf("cannot extract source from the loginput state: %w", err)
+		}
+		source, ok := sourceValue.(string)
+		if !ok {
+			return true, nil
+		}
+
+		var filestreamID string
+		for id, matcher := range matchers {
+			if matcher(source) {
+				filestreamID = id
+				break
+			}
+		}
+		if filestreamID == "" {
+			return true, nil
+		}
+
+		newKey := loginputToFilestreamKey(key, filestreamID)
+		log.Infof("found loginput state `%s` to take over by `%s`", key, newKey)
+
+		newState := loginputToFilestream(state)
+
+		toSet[newKey] = newState
+		toRemove[key] = struct{}{}
+
+		return true, nil
+	})
+
+	return toSet, toRemove, err
+}
+
+// findFilestreams finds filestream inputs that are marked as `take_over: true`
+// and creates a file matcher for each such filestream for the future use in state
+// processing
+func findFilestreams(log *logp.Logger, cfg *cfg.Config) (matchers filestreamMatchers, err error) {
+	matchers = make(filestreamMatchers)
+
+	for _, input := range cfg.Inputs {
+		inputCfg := defaultInputConfig()
+		err := input.Unpack(&inputCfg)
+		if err != nil {
+			return nil, fmt.Errorf("failed to unpack input configuration: %w", err)
+		}
+		if inputCfg.Type != "filestream" || !inputCfg.TakeOver {
+			continue
+		}
+		if _, exists := matchers[inputCfg.ID]; exists || inputCfg.ID == "" {
+			return matchers, fmt.Errorf("filestream `%s` in `take over` mode requires a unique ID", inputCfg.ID)
```

Comment:
```suggestion
			return matchers, fmt.Errorf("filestream with ID `%s` in `take over` mode requires a unique ID. Add the `id:` key with a unique value.", inputCfg.ID)
```

I think the error here will be clearer if we specify explicitly that the string we are printing is the `id` field we parsed from the config, particularly when the value is the empty string.

It would also be nice to try to suggest the exact correction to make in adding an `id:` key and value.

This is a nice way to force people to add unique IDs to their filestream inputs 👍